src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,176 @@
+        }
+        if (list.size() == 0) {
+            return;
+        }
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (value == null) {
+            throw new IllegalArgumentException("Cannot set the value because it is null");
+        }
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> currentNode = findNodeByIndex(index);
+        T oldValue = currentNode.value;
+        unlink(currentNode);
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        if (object == null) {
+            while (currentNode != null) {
+                if (currentNode.value == null) {
+                    unlink(currentNode);
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        } else {
+            while (currentNode != null) {
+                if (object.equals(currentNode.value)) {
+                    unlink(currentNode);
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        Node<T> prev = node.prev;
+        Node<T> next = node.next;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            node.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            node.next = null;
+        }
+        node.value = null;
+        size--;
+    }
+
+    private void indexValidation(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> currentNode = first;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        return currentNode;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
```

Comment:
```suggestion
        Node(Node<T> prev, T value, Node<T> next) {
```

```suggestion
        private Node(Node<T> prev, T value, Node<T> next) {
```
it has no effect, so better be upfront about it, to not confuse anyone